src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.ComposeCache.targets

```diff
@@ -0,0 +1,370 @@
+      <_RuntimeRefDir>$([System.IO.Path]::Combine($(_CrossProjFileDir), "runtimeref"))</_RuntimeRefDir> <!-- flat app managed assemblies -->
+      <_NetCoreRefDir>$([System.IO.Path]::Combine($(_CrossProjFileDir), "netcoreapp"))</_NetCoreRefDir> <!-- flat netcore app assemblies-->
+      <_RuntimeOptimizedDir>$([System.IO.Path]::Combine($(_CrossProjFileDir), "runtimopt"))</_RuntimeOptimizedDir> <!-- optimized app managed assemblies in nuget cache layout -->
+      <_CrossProjFile>$([System.IO.Path]::Combine($(_CrossProjFileDir), "Optimize_Crossgen_$(RuntimeIdentifier)_$(TargetFramework)_$(FX_Version).csproj"))</_CrossProjFile>
+      <_CrossProjAssetsFile>$([System.IO.Path]::Combine($(_CrossProjFileDir), "obj", project.assets.json))</_CrossProjAssetsFile>
+      <_tfm>$(TargetFrameworkMoniker)</_tfm>
+    </PropertyGroup>
+
+    <MakeDir  Directories="$(_CrossProjFileDir)"/>
+    <MakeDir  Directories="$(_RuntimeDir)"/>
+    <WriteLinesToFile
+        File="$(_CrossProjFile)"
+        Lines="$(_CrossgenProjContent)"
+        Overwrite="true" />
+  </Target>
+
+  <!--
+    ============================================================
+                                        _SetupCrossgenStage
+    ============================================================
+    -->
+  <Target Name="_SetupCrossgenStage">
+    <!-- Copy managed files to temp directory maintaining the structure -->
+    <Copy SourceFiles = "@(_ManagedResolvedFileToPublishCandidates)"
+          DestinationFiles="$(_RuntimeDir)\%(_ManagedResolvedFileToPublishCandidates.DestinationSubPath)"
+          OverwriteReadOnlyFiles="$(OverwriteReadOnlyFiles)"
+          Retries="$(CopyRetryCount)"
+          RetryDelayMilliseconds="$(CopyRetryDelayMilliseconds)"
+          UseHardlinksIfPossible="$(CreateHardLinksForPublishFilesIfPossible)"
+          UseSymboliclinksIfPossible="$(CreateSymbolicLinksForPublishFilesIfPossible)">
+
+      <Output TaskParameter="DestinationFiles" ItemName="FileWrites"/>
+    </Copy>
+
+    <!-- Copy managed files to  a flat temp directory for passing it as ref -->
+    <Copy SourceFiles = "@(_ManagedResolvedFileToPublishCandidates)"
+          DestinationFolder="$(_RuntimeRefDir)"
+          OverwriteReadOnlyFiles="$(OverwriteReadOnlyFiles)"
+          Retries="$(CopyRetryCount)"
+          RetryDelayMilliseconds="$(CopyRetryDelayMilliseconds)"
+          UseHardlinksIfPossible="$(CreateHardLinksForPublishFilesIfPossible)"
+          UseSymboliclinksIfPossible="$(CreateSymbolicLinksForPublishFilesIfPossible)">
+
+      <Output TaskParameter="DestinationFiles" ItemName="FileWrites"/>
+    </Copy>
+
+  </Target>
+  <!--
+    ============================================================
+                                        _RestoreCrossgen
+Restores netcoreapp and pulbishes it to a temp directory
+    ============================================================
+    -->
+  <Target Name="_RestoreCrossgen">
+    <!-- Restore Crossgen-->
```

Comment:
(nit) this comment doesn't really add value since it is just repeating the name of the target.